opentelemetry-api/src/opentelemetry/util/types.py

```diff
@@ -13,17 +13,28 @@
 # limitations under the License.
 
 
-from typing import Callable, Mapping, Optional, Sequence, Union
+from typing import Callable, Mapping, Optional, Sequence, Tuple, Union
 
 AttributeValue = Union[
     str,
     bool,
     int,
     float,
-    Sequence[Union[None, str]],
-    Sequence[Union[None, bool]],
-    Sequence[Union[None, int]],
-    Sequence[Union[None, float]],
+    Sequence[Optional[str]],
+    Sequence[Optional[bool]],
+    Sequence[Optional[int]],
+    Sequence[Optional[float]],
+]
+AttributeValueAsKey = Union[
+    str,
+    bool,
+    int,
+    float,
+    Tuple[Optional[str], ...],
+    Tuple[Optional[bool], ...],
+    Tuple[Optional[int], ...],
+    Tuple[Optional[float], ...],
 ]
 Attributes = Optional[Mapping[str, AttributeValue]]
+AttributesAsKey = Tuple[Tuple[str, AttributeValueAsKey], ...]
```

Comment:
I looked up [the Ellipsis](https://stackoverflow.com/questions/772124/what-does-the-ellipsis-object-do) here but I'm still unsure as to why it's needed here. If the key and value are already in the first element of the tuple, what goes in the rest of the tuple?

Reply:
It is the way to type a variable length tuple

https://docs.python.org/3/library/typing.html#typing.Tuple:~:text=To%20specify%20a%20variable%2Dlength%20tuple%20of%20homogeneous%20type%2C%20use%20literal%20ellipsis%2C%20e.g.%20Tuple%5Bint%2C%20...%5D.